File: src/main/java/com/flowcommerce/core/domain/member/Commission.java
package com.flowcommerce.core.domain.member;

import com.flowcommerce.core.support.error.CoreException;
import com.flowcommerce.core.support.error.ErrorType;

import java.math.BigDecimal;

public record Commission(BigDecimal rate) {
    public Commission{
        if (rate == null) {
            throw new CoreException(ErrorType.COMMISSION_REQUIRED);
        }
        if (rate.compareTo(BigDecimal.ZERO) < 0) {
            throw new CoreException(ErrorType.COMMISSION_NEGATIVE);
        }
        if (rate.compareTo(BigDecimal.ONE) > 0) {
            throw new CoreException(ErrorType.COMMISSION_EXCEEDED);
        }
    }

    public BigDecimal calculate(BigDecimal amount) {
        return amount.multiply(rate);
    }
}

File: src/main/kotlin/com/flowcommerce/core/domain/member/Commission.kt
package com.flowcommerce.core.domain.member

import com.flowcommerce.core.support.error.CoreException
import com.flowcommerce.core.support.error.ErrorType
import java.math.BigDecimal

data class Commission(val rate: BigDecimal) {
    init {
        if (rate < BigDecimal.ZERO) {
            throw CoreException(ErrorType.COMMISSION_NEGATIVE)
        }
        if (rate > BigDecimal.ONE) {
            throw CoreException(ErrorType.COMMISSION_EXCEEDED)
        }
    }

    fun calculate(amount: BigDecimal): BigDecimal = amount.multiply(rate)
}
